rfc/0001-oauth-access-tokens.rst

```diff
@@ -0,0 +1,345 @@
+.. header:: This proposal is `discussed at this pull request <https://github.com/hail-is/hail-rfc/pull/0>`_.
+            **After creating the pull request, edit this file again, update the
+            number in the link, and delete this bold sentence.**
+.. sectnum::
+.. contents::
+.. role:: python(code)
+
+Motivation
+==========
+
+This proposal focuses on the way by which users of Hail Batch and Hail Query-on-Batch
+(from now on referred to as the Hail Service) authorize programmatic access to the hail API.
+
+The Hail Service authenticates users using the OAuth2 protocol, relying on either
+GCP IAM or Azure AD as the identity providers. However, while the Hail Service
+relies on these identity providers for authentication, it currently does *not* use them
+to authorize access to Hail APIs. The Hail ``auth`` service acts as an Authorization
+Server for the Hail API, minting long-lived tokens after the OAuth2 flow that are persisted
+on user machines. Minting our own tokens imposes a maintenance and security burden
+on the Hail team and any operators of a Hail Service.
+
+This proposal deprecates the use of hail-minted tokens in favor of using
+access tokens from the identity providers listed above to authorize API access.
+This removes the security burden of minting and protecting our own authorization
+tokens while reducing code complexity since cloud access tokens are already
+used within the hail codebase to access cloud APIs.
+
+Proposed Change Specification
+=============================
+
+Currently, requests to the Hail APIs send a hail-minted Bearer token in the
+``Authorization`` header of HTTP requests. This token is stored in a well-known
+location on the user's disk.
+For user machines, this file is persisted during the login flow ``hailctl auth login``.
+For use in Batch jobs, the tokens are stored in Kubernetes secrets and delivered
+to the Batch Worker as part of the job spec.
+
+This proposal adds the ability for HTTP requests from hail clients to send
+OAuth2 access tokens in the ``Authorization`` header instead of hail-minted
+tokens. The ``auth`` service will:
+
+- Assert the validity, expiration and audience of access tokens and associate
+  them with users of the system.
+- Support hail-minted bearer tokens for backwards compatibility with old clients.
+
+Hail clients will be updated to use access tokens in requests to Hail APIs. How
+they do so is described in the following subsections.
+
+
+Overview of Relevant OAuth2 Background
+--------------------------------------
+
+Prior to discussing the details of the implementation, it is worth covering some
+background on OAuth2. Note that much of this functionality is encapsulated in the
+Google OAuth and AAD client libraries that we use, but a thorough understanding
+is valuable to ensure that we are using them properly.
+
+We'll consider four primary entities in an OAuth2 interaction:
+
+- The user/identity
+- The client (e.g. the hail python library)
```

Comment:
I split those out because the authentication process is slightly different (interactive flow vs just key/password submission from the filesystem), but I can cut that out if you don't think that's important in this discussion.